### PR TITLE
dev_multiple_types:  Support multiple buffer types in one platform

### PR DESCRIPTION
**Pull request by mcmillan03**
Basically merging reference and quantized_reference platforms.  And refactoring all the other platforms to deal with the fallout from the changes at higher levels of the framework architecture.

### CODE REVIEW

**Comment by mcmillan03** [comment]
The main question is whether or not reference_combined should replace the reference platform (and delete quantized_reference).

**Comment by nicholaiTukanov** [comment]
> The main question is whether or not reference_combined should replace the reference platform (and delete quantized_reference).

I think it should so that we don't have duplicate logic in the code base.

File: docs/Defining_Platforms.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    static uint32_t const   NUM_FMA{FLOAT_NUM_FMA};
+    static uint32_t const   NUM_MAX{FLOAT_NUM_MAX};
+    static uint32_t const  NUM_LOAD{FLOAT_NUM_LOAD};
+    static uint32_t const NUM_STORE{FLOAT_NUM_STORE};
+};
+```
+
+Note that the prefix must be removed so that each `<Type>Buffer` class defines exactly the same set of constants.
+
+TBD: make sure these are compile-time constants.
```
Comment (nicholaiTukanov) [inline docs/Defining_Platforms.md:116]: doesn't `const` already ensure that?

File: docs/Defining_Platforms.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+* Copy assignment operator
+* Move assignment operator
+* Destructor (not virtual)
+* `size_t size() const;` - return the number of elements in the buffer
+* `value_type [const] *data() [const];` - return pointer to beginning of buffer memory
+* `value_type [const] &operator[](size_t index) [const];` - index into array using square bracket.
+* `void swap(<Type>Buffer &other);` - swap the contents of this buffer with `other`.
+* `accum_type zero() const;' - mainly used by quantized types to report the bit representation of a zero value.
+
+### Microkernels
```
Comment (nicholaiTukanov) [inline docs/Defining_Platforms.md:149]: We are going to need more details on how to write the microkernels somewhere

File: docs/Defining_Platforms.md (hunk 1 above)
Comment (mcmillan03) [reply docs/Defining_Platforms.md:116]: Need to confirm either experimentally with Compiler Explorer or relevant language in the ISO C++ Specification.

File: docs/Defining_Platforms.md (hunk 2 above)
Comment (mcmillan03) [reply docs/Defining_Platforms.md:149]: Need to get Upasana to document each microkernel.

**Review by sei-smcmillan — APPROVED** [review]
Changes look good.  Noting that unit_tests, demos and other applications are not really supported for quantized types (only test_buffer.cpp).  Deferring quantized testing for another PR.

**Review by sei-smcmillan — APPROVED** [review]

**Pull request merged**